src/source/parserutil.d

```diff
@@ -8,26 +8,35 @@ auto match(Lexer, TokenType)(ref Lexer lexer, TokenType type) {
 		return token;
 	}
 
-	import std.conv, std.string;
-	auto error = token.type == TokenType.Invalid
-		? token.error.toString(lexer.context)
-		: format!"expected '%s', got '%s'."(to!string(type),
-		                                    to!string(token.type));
-
-	import source.exception;
-	throw new CompileException(token.location, error);
+	import std.conv;
+	throw unexpectedTokenError(lexer, to!string(type));
 }
 
 auto unexpectedTokenError(Lexer)(ref Lexer lexer, string expected) {
 	auto token = lexer.front;
 
-	import std.conv, std.string;
-	auto error = token.type == Lexer.TokenType.Invalid
-		? token.error.toString(lexer.context)
-		: format!"expected %s, not `%s`."(expected,
-		                                  token.toString(lexer.context));
+	import std.format;
+	string error;
+	switch (token.type) {
+		case Lexer.TokenType.Invalid:
+			error = token.error.toString(lexer.context);
+			break;
+
+		case Lexer.TokenType.End:
+			error = format!"expected %s, not the end of input."(expected);
```

Comment:
You can check for `isMixin` and chose `end of input` or `end of file`. The lexer already does this here: https://github.com/snazzy-d/sdc/blob/a9341c219fb25c20244d8fd78aaef7f8f9224f4d/src/source/lexbase.d#L222-L224